src/analysis/special_functions/trigonometric.lean

```diff
@@ -2205,6 +2251,47 @@ begin
     ring },
 end
 
+lemma tan_add {x y : ℂ}
+  (h : ((∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) ∧ ∀ l : ℤ, y ≠ (2 * l + 1) * π / 2)
+     ∨ ((∃ k : ℤ, x = (2 * k + 1) * π / 2) ∧ ∃ l : ℤ, y = (2 * l + 1) * π / 2)) :
+  tan (x + y) = (tan x + tan y) / (1 - tan x * tan y) :=
+begin
+  cases h,
+  { rw [tan, sin_add, cos_add,
+        ← div_div_div_cancel_right (sin x * cos y + cos x * sin y)
+            (mul_ne_zero (cos_ne_zero_iff.mpr h.1) (cos_ne_zero_iff.mpr h.2)),
+        add_div, sub_div],
+    simp only [←div_mul_div, ←tan, mul_one, one_mul,
+              div_self (cos_ne_zero_iff.mpr h.right), div_self (cos_ne_zero_iff.mpr h.left)] },
+  { cases h.1 with k h1, cases h.2 with l h2,
+    have hx := tan_int_mul_pi_div_two (2*k+1),
+    have hy := tan_int_mul_pi_div_two (2*l+1),
+    have hxy := tan_int_mul_pi_div_two (2*k+1+(2*l+1)),
+    simp only [int.cast_add, int.cast_bit0, int.cast_mul, int.cast_one, ← h1, ← h2] at hx hy hxy,
```

Comment:
I know there's a way to golf these four lines but I can't seem to make it work. Perhaps someone more experienced than I can enlighten me.

Reply:
Assuming you use the `rcases` line suggested above, you can do this case of the proof using
```lean
{ rw [show tan ((2 * k + 1) * π / 2) = 0, by exact_mod_cast tan_int_mul_pi_div_two (2*k+1),
        show tan ((2 * l + 1) * π / 2) = 0, by exact_mod_cast tan_int_mul_pi_div_two (2*l+1), 
        add_zero, zero_div, tan_eq_zero_iff],
    use 2*k + 2*l + 2,
    field_simp, 
    ring }
```
However the `exact_mod_cast` are *very* slow (@robertylewis).

Reply:
This looks like bad behavior in the heuristic splitting stage of `norm_cast`. I'm not completely sure what's triggering it. Feel free to open an issue but I can't promise when I'll have time to look at it.